### DISCUSSION TITLE
Sound?

### DISCUSSION THREAD

**Discussion by samster395** [gab-k/RP2040-Decoder#7] [Ideas]
Love what you've done.

I was wondering if you think it would be possible to make a version of this with sound output?

**Comment by gab-k** [comment]
I've been thinking about that too.  It is certainly possible, but this would involve both relatively extensive hardware and software changes.

**Reply by samster395** [reply]
Yeah, I thought it might be possible but one thing I did find that might be an issue is apparently it's hard to make sounds play over the top of each other which conventional decoders can do.

I'm kinda out of my depth when it comes to this kinda stuff tbh, just wondered if it was possible and would love to see it.

**Reply by gab-k** [reply]
By playing sounds on top of each other do you mean, for example, that a sound is already being played and another sound is to be played simultaneously during playback?
This should be realizable in software.

**Reply by samster395** [reply]
Yeah like say the engine running sound is playing and then you play the horn on top of it.

I was looking at some thread about someone trying to do with an Arduino chip or something similar and the replies were saying there's not enough room in the buffer to do that and some other issues.

I can't seem to find the thread now.

Like I said im outta my depth when it comes to this but it's cool to hear you think it's possible with the RP2040, I guess it's because it is a more powerful microcontroller.

**Reply by questor** [reply]
first I have to admit that I never heard up to now really how a sound-decoder sounds. I also have up to now never experiences such an decoder so that I could generate ideas how they work.

the first point for sound-decoders is how to store the samples (storing 5s of mono 44khz, 16bit samples takes uncompressed 5*44000*2 bytes = 440kB). I would start with this:
- some samples need to be stored directly, but compressed (horn) (for example with this: https://phoboslab.org/log/2023/02/qoa-time-domain-audio-compression and with 32khz)
- some other samples (engine sound) I would store either as samples directly (steam-engine sounds) or generate them dynamically (diesel engine sound, eletric engine sound) and adapt them based on the speed of the locomotive.

so first question would be: how much space for samples is available (and maybe optimize the firmware a bit to take less space) and how  much processing power is left for the sound engine.

btw, mixing two audio signals is rather easy, it's done by adding both samples...

some links how to generate sounds are https://github.com/DasEtwas/enginesound and https://github.com/TheDIYGuy999/Rc_Engine_Sound_ESP32, but I have not taken a deeper look if they are applicable for trains and for this decoder...

**Reply by gab-k** [reply]
Flash size is 64MBit and most of it should be unused. Considering there are probably not exactly hi-fi demands on the sound I dont think even 16Bits is needed. The Idea of generating sound dynamically seems to be very interesting however. In terms of processing power and software optimization there is still a lot of room for improvement.

**Reply by questor** [reply]
after looking around if I remember correctly Märklin is using 16bit samples, but only ~16kHz (analog phone line quality). and both esu and märklin seem not to use procedural generation but samples (and maybe processed based on speed). and sadly both seem to use encrypted samples so no easy testing possible with their samples.

**Reply by samster395** [reply]
> after looking around if I remember correctly Märklin is using 16bit samples, but only ~16kHz (analog phone line quality). and both esu and märklin seem not to use procedural generation but samples (and maybe processed based on speed). and sadly both seem to use encrypted samples so no easy testing possible with their samples.

I found these samples for DCC sound, they might be usable, there is sounds from Japanese and German trains.
I believe these are supposed to be used with the ESU LOKSOUND decoders but might be able to be used for testing.

https://desktopstation.net/sounds/index_eng.html

**Reply by questor** [reply]
what are some speakers I can order in germany to do tests? and where can I connect them, on the AUX? or GPIO (which could use PWM, right)?

**Reply by gab-k** [reply]
Yes, all the GPIOs are PWM capable. You can even configure them using CVs or you can directly modify the source code which might be easier to be honest. You probably need to amplify your filtered PWM signal.

**Reply by questor** [reply]
I've found this: https://www.austromodell.at/index.php?site=webshop&kategorie=152&anr=680&detail=1&farbe=  which looks suitable, but to be honest I have no idea what I need to be able to connect the speaker to the decoder. assuming for the first tests I don't need an amplifier, what do I need to do to connect the speaker with 8ohm and 1W to the decoder? (of course I will come back with questions if I DO need to amplifiy the sound ;) )

**Reply by gab-k** [reply]
> I've found this: https://www.austromodell.at/index.php?site=webshop&kategorie=152&anr=680&detail=1&farbe= which looks suitable, but to be honest I have no idea what I need to be able to connect the speaker to the decoder.

Can't really say much about the speaker to be honest.

> assuming for the first tests I don't need an amplifier, what do I need to do to connect the speaker with 8ohm and 1W to the decoder? (of course I will come back with questions if I DO need to amplifiy the sound ;) )

You will not get 1 W out of the RP2040, directly loading the controller with the rather low impedance speaker will not work, i suggest reading Chapter 3.4 of [this](https://datasheets.raspberrypi.com/rp2040/hardware-design-with-rp2040.pdf), which provides a good introduction into the topic and is a short read. If you delve deeper into the topic you will notice that there is some other things to worry about aswell, a clean power supply for example, provided non distorted sound is a priority. Also notice that there is also the option to use an external DAC IC as described in 3.4.2. which will probably the easier way to implement something like this.

**Reply by ABaboshin** [reply]
Maybe it is outdated, but I just came across because I'm about to develop my own dcc decoder based on rp2040 too :) and so I was able to google this amazing project.

I cannot say much about exactly this speaker, because there is no documentation on the linked website. But in a case of a regular speaker (which you can get for ~1€ or less) it will be red for +, black for -. Which means: + on a desired gpio for pwm, - on zero.

In case with an amplifier it will be the same: 
gpio pwm - amplifier l/r
gnd - amplifier gnd
and then
amplifier r or l + - speaker red
amplifier r or l - - speaker black

For the test it will be even enough to use a regular headphone and connect pwm pin/amplifier output to l or r and ground correspondently.

In case of a cheap speaker you will probably need to build low pass / high pass filters before the speaker.

if it is about howto play an audio, the pico arduino sources are usable and have good examples.

**Comment by LaurentDelaon** [comment]
Hi Gab-k,
congratulation for your beautiful project! 
I am not DCC specialist nor rp2040...
I would use MP3 DFPlayer Player Module with your decoder. Seem the simplest way to add sound.
With powerful possibilities.

Questions:
a) how obtain a RxTx on rp2040 on two GPIO ?
b) what'sCV numbers use to activate some sounds according to DCC norm ? 
  1-a whristle short
  2-a whristle long
  3-sound vapor movement as  repetitive read may have a variable frequency  on  CV's speed.
  4-and so on...

 (out of scope) suggest to add dot/plus orientation on pcb.
I want realise decoder on JLCPCB but unable to put BOM or centroid list ...

regards.
Laurent.

**Reply by gab-k** [reply]
Hi Laurent,

In order to realize a UART communication I would recommend you to look at the [Raspberry Pi Pico uart examples ](https://github.com/raspberrypi/pico-examples/tree/master/uart). A simple UART communication should be relatively easy to implement, especially since you only have to send commands. The hello_uart example shows the simplest implementation, but you should keep in mind that the UART function is best called from core0, since the control algorithm runs on core1, which is time-critical.

Gabriel

**Reply by Gibbo97** [reply]
Hi Laurent. I also was unable to put together the BOM. The parts were not being... auto detected? for lack of a better term, on JLCPCB. I would have to know much more about the parts and parameters required. I'm only a software engineer, never hardware so out of my wheelhouse unfortunately. I have had some boards printed for another project but they didn't require any surface mount parts (and the files were all ready to be uploaded into JLCPCB), it seems there is considerable additional complexity in producing these files.

If you are still interested, there is now someone selling some slightly modified boards put together already here: https://ilabs.se/product/opendec02/

**Reply by questor** [reply]
Hi! I also had issues with the BOM (I'm also more a software-guy than hardware), but with the excellent help from Gabriel I was able to order fully assembled boards. Because I also had issues Gabriel provided some additional files here: https://github.com/gab-k/RP2040-Decoder/issues/6   hope that helps. -Kai

**Comment by LaurentDelaon** [comment]
HI Gabriel,

thank's for your reply.
I look at your link.
I well know teensyduino, but I code them via IDE which seem a bit more simple.

I  know regulation , normaly you should use a time sample of around 0.1s-0.25s for PID regulator.
Due to lag you may have an othershoot. 
As me no need to feedforward in this case of speed control...
Have you ever see a train with constante speed when it climbed ? seem not ;o)...
I don't have any dcc train so maybe I'am wrong...

thank's a lot.

Laurent.

----- Mail original -----
De: "Gabriel Koppenstein" ***@***.***>
À: "gab-k/RP2040-Decoder" ***@***.***>
Cc: "LaurentDelaon" ***@***.***>, "Comment" ***@***.***>
Envoyé: Lundi 14 Août 2023 21:00:25
Objet: Re: [gab-k/RP2040-Decoder] Sound? (Discussion #7)





Hi Laurent, 

In order to realize a UART communication I would recommend you to look at the Raspberry Pi Pico uart examples . A simple UART communication should be relatively easy to implement, especially since you only have to send commands. The hello_uart example shows the simplest implementation, but you should keep in mind that the UART function is best called from core0, since the control algorithm runs on core1, which is time-critical. 

Gabriel
